internal/controllers/cluster/cluster_controller_phases.go

```diff
@@ -143,132 +141,149 @@ func (r *Reconciler) reconcileExternal(ctx context.Context, cluster *clusterv1.C
+		if s.cluster.DeletionTimestamp.IsZero() {
+			cluster.Status.InfrastructureReady = true
+			conditions.MarkTrue(cluster, clusterv1.InfrastructureReadyCondition)
+		}
 		return ctrl.Result{}, nil
 	}
 
 	// Call generic external reconciler.
-	infraReconcileResult, err := r.reconcileExternal(ctx, cluster, cluster.Spec.InfrastructureRef)
+	obj, err := r.reconcileExternal(ctx, cluster, cluster.Spec.InfrastructureRef)
 	if err != nil {
-		return ctrl.Result{}, err
-	}
-	// Return early if we need to requeue.
-	if infraReconcileResult.RequeueAfter > 0 {
-		return ctrl.Result{RequeueAfter: infraReconcileResult.RequeueAfter}, nil
-	}
-	// If the external object is paused, return without any further processing.
-	if infraReconcileResult.Paused {
-		return ctrl.Result{}, nil
-	}
-	infraConfig := infraReconcileResult.Result
+		if apierrors.IsNotFound(err) {
+			s.infraClusterIsNotFound = true
 
-	// There's no need to go any further if the Cluster is marked for deletion.
-	if !infraConfig.GetDeletionTimestamp().IsZero() {
-		return ctrl.Result{}, nil
+			if !cluster.DeletionTimestamp.IsZero() {
+				// Tolerate infra cluster not found when the cluster is being deleted.
+				return ctrl.Result{}, nil
+			}
+
+			if cluster.Status.InfrastructureReady {
+				// Infra object went missing after the cluster was up and running
+				return ctrl.Result{}, reconcile.TerminalError(errors.Errorf("%s has been deleted after being ready", cluster.Spec.InfrastructureRef.Kind))
+			}
+			log.Info(fmt.Sprintf("Could not find %s, requeuing", cluster.Spec.InfrastructureRef.Kind))
+			return ctrl.Result{RequeueAfter: externalReadyWait}, nil
+		}
+		return ctrl.Result{}, err
 	}
+	s.infraCluster = obj
 
 	// Determine if the infrastructure provider is ready.
-	preReconcileInfrastructureReady := cluster.Status.InfrastructureReady
-	ready, err := external.IsReady(infraConfig)
+	ready, err := external.IsReady(s.infraCluster)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
-	cluster.Status.InfrastructureReady = ready
-	// Only record the event if the status has changed
-	if preReconcileInfrastructureReady != cluster.Status.InfrastructureReady {
-		r.recorder.Eventf(cluster, corev1.EventTypeNormal, "InfrastructureReady", "Cluster %s InfrastructureReady is now %t", cluster.Name, cluster.Status.InfrastructureReady)
+	if ready && !cluster.Status.InfrastructureReady {
+		log.Info("Infrastructure provider has completed cluster infrastructure provisioning and reports status.ready", cluster.Spec.InfrastructureRef.Kind, klog.KObj(s.infraCluster))
 	}
 
 	// Report a summary of current status of the infrastructure object defined for this cluster.
 	conditions.SetMirror(cluster, clusterv1.InfrastructureReadyCondition,
-		conditions.UnstructuredGetter(infraConfig),
+		conditions.UnstructuredGetter(s.infraCluster),
 		conditions.WithFallbackValue(ready, clusterv1.WaitingForInfrastructureFallbackReason, clusterv1.ConditionSeverityInfo, ""),
 	)
```

Comment:
I think we also should have an if/else for fallBack like we have in machine_controller_phases.go